### PR TITLE
[WIP][KYUUBI #2238] Support Flink 1.15

### PR DESCRIPTION
**Pull request by wolfboys**
### Why are the changes needed?

Flink 1.15.0 is out, Kyuubi needs to support it.

Please clarify why the changes are needed. For instance

1. new add flink 1.15 support
2. new add `flink-1.14` and `flink-1.15` maven profiles to support both Flink 1.14 & 1.15 

flink 1.14  all functions and codes are kept unchanged without any impact

flink 1.15 support part of the work is still in progress, At present, 32 test cases have passed and 5 have failed, You can review, put forward suggestions for improvement, I will continue to improve.

### how to build with flink 1.14

```
mvn clean install -DskipTests -Pflink-1.14
```

### how to build with flink 1.15

```
mvn clean install -DskipTests -Pflink-1.15
```

### how to run test cases with flink 1.14

enable `flink-1.14` maven profile  and run the test cases

### how to run test cases with flink 1.15

enable `flink-1.15` maven profile  and run the test cases

### _How was this patch tested?_

- [x]  Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1330,19 +1331,19 @@
 
             <dependency>
                 <groupId>org.apache.flink</groupId>
-                <artifactId>flink-yarn_${scala.binary.version}</artifactId>
+                <artifactId>flink-yarn${flink.module.scala.suffix}</artifactId>
                 <version>${flink.version}</version>
             </dependency>
```
Comment (pan3793) [inline pom.xml:1336]: This dependency seems does use anymore, we can remove it.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala [hunk 2]
```diff
@@ -142,7 +151,18 @@ class ExecuteStatement(
           case TypedResult.ResultType.PAYLOAD =>
             (1 to result.getPayload).foreach { page =>
               if (rows.size < resultMaxRows) {
-                rows ++= executor.retrieveResultPage(resultId, page).asScala
+                val result = executor.retrieveResultPage(resultId, page).asScala.toList
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:154]: Would you please add comments here? e.g. FLINK-xxx changes the return type from xx to yy

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala [hunk 3]
```diff
@@ -142,7 +151,18 @@ class ExecuteStatement(
...
+                result.headOption match {
+                  case None =>
+                  case Some(r) =>
+                    // for flink 1.14
+                    if (r.getClass == classOf[Row]) {
+                      rows ++= result.asInstanceOf[List[Row]]
+                    } else {
+                      // for flink 1.15+
+                      rows ++= result.map(r => convertToRow(r.asInstanceOf[RowData], dataTypes))
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:163]: From my understanding, `RowData` is more future proof than `Row`, then `convertToDataRow` should more reasonable?
@link3280 @SteNicholas

Comment (link3280) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:164]: IIUC, `RowData` is Flink's internal data type, thus I think `Row` is more suitable for our case.

Comment (pan3793) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:164]: Are we suffering from [FLINK-21456](https://issues.apache.org/jira/browse/FLINK-21456)?

**Comment by wolfboys** [comment]
@SteNicholas @deadwind4 PTAL. thx

**Review by pan3793 — APPROVED** [review]

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala (hunk 3 above)
Comment (deadwind4) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:164]: I talk with Cheng Pan offline, LGTM.

**Review by deadwind4 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merging to master

**Pull request closed without merging**